internal/pkg/service/appsproxy/proxy/apphandler/authproxy/config.go

```diff
@@ -39,10 +40,20 @@ func (m *Manager) proxyConfig(app api.AppConfig, authProvider provider.Provider,
 
 	// Render the selector page, if login is needed, it is not an internal URL
 	v.OnNeedsLogin = func(w http.ResponseWriter, req *http.Request) (stop bool) {
-		if !strings.HasPrefix(v.ProxyPrefix, req.URL.Path) && m.providerSelector.providerIDFromCookie(req) != "" {
-			m.providerSelector.clearCookie(w, req)
-			w.Header().Set("Location", req.URL.Path)
-			w.WriteHeader(http.StatusFound)
+		if !strings.HasPrefix(req.URL.Path, v.ProxyPrefix) && m.providerSelector.providerIDFromCookie(req) != "" {
+			if selector, ok := req.Context().Value(selectorHandlerCtxKey).(*SelectorForAppRule); ok {
+				// Go back and render the selector page
+				req = req.WithContext(context.WithValue(req.Context(), ignoreProviderCookieCtxKey, true))
+				if err := selector.ServeHTTPOrError(w, req); err != nil {
+					m.pageWriter.WriteError(w, req, &app, err)
+				}
+			} else {
+				// Fallback, it shouldn't happen.
+				// Clear the cookie and redirect to the same path, so the selector page is rendered.
+				m.providerSelector.clearCookie(w, req)
+				w.Header().Set("Location", req.URL.Path)
+				w.WriteHeader(http.StatusFound)
+			}
```

Comment:
See updated test bellow.

Originally, I had only the `else` branch here.

Both branches works, but `else` branch generates one more redirect, and @jachym-tousek-keboola don't expect it, in the test.